### PR TITLE
add validation for upgrade schedule annotations

### PR DESCRIPTION
**Pull request by anvddriesch**
Towards https://github.com/giantswarm/giantswarm/issues/18077

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,7 +6,7 @@ require (
 	github.com/blang/semver/v4 v4.0.0
 	github.com/dylanmei/iso8601 v0.1.0
 	github.com/dyson/certman v0.2.1
-	github.com/giantswarm/apiextensions/v3 v3.32.0
+	github.com/giantswarm/apiextensions/v3 v3.32.1-0.20210908101333-093943a90ae0
```
Comment (njuettner) [inline go.mod:9]: this looks like a branch

File: pkg/aws/v1alpha3/cluster/validate_cluster.go [hunk 2]
```diff
@@ -119,6 +143,75 @@ func (v *Validator) ValidateUpdate(request *admissionv1.AdmissionRequest) (bool,
...
+					annotation.UpdateScheduleTargetTime,
+					updateTime),
+			)
+		}
+	}
+	return nil
+}
+
+func UpgradeScheduleTimeIsValid(updateTime string) bool {
+	// parse time
```
Comment (njuettner) [inline pkg/aws/v1alpha3/cluster/validate_cluster.go:160]: I would change the description to something useful like how the time should look like

File: pkg/aws/v1alpha3/cluster/validate_cluster.go [hunk 3]
```diff
@@ -119,6 +143,75 @@ func (v *Validator) ValidateUpdate(request *admissionv1.AdmissionRequest) (bool,
...
+	t, err := time.Parse(time.RFC822, updateTime)
+	if err != nil {
+		return false
+	}
+	//time already passed
+	if t.Before(time.Now()) {
+		return false
+	}
+	// time too far in the future
+	if time.Until(t) > 4380*time.Hour {
```
Comment (njuettner) [inline pkg/aws/v1alpha3/cluster/validate_cluster.go:170]: what is the reason of 4380 hours? maybe add this to the comment

**Review by njuettner — COMMENTED** [review]
the admission controller runs in several time zones, which default time are we using if someone applies it? which is used for comparison?

File: go.mod (hunk 1 above)
Comment (anvddriesch) [reply go.mod:9]: it's master since we did not release apiextensions yet :)

File: pkg/aws/v1alpha3/cluster/validate_cluster.go (hunk 3 above)
Comment (anvddriesch) [reply pkg/aws/v1alpha3/cluster/validate_cluster.go:170]: it's 6 months. And yeah, will add a comment.

File: go.mod (hunk 1 above)
Comment (njuettner) [reply go.mod:9]: then what about releasing the apiextensions and update this here :) ?

File: pkg/aws/v1alpha3/cluster/validate_cluster.go (hunk 2 above)
Comment (anvddriesch) [reply pkg/aws/v1alpha3/cluster/validate_cluster.go:163]: I added an example!

File: go.mod (hunk 1 above)
Comment (anvddriesch) [reply go.mod:9]: that's the plan. We also need to do that in the operator.

**Comment by anvddriesch** [comment]
Tested this quite a bit and it seems fine.

**Review by njuettner — APPROVED** [review]

**Pull request merged**